backend/Application/Interop/InteropBinding.cs

```diff
@@ -0,0 +1,131 @@
+            var ptrToStringAnsi = Marshal.PtrToStringAnsi(result);
+            return new InteropResult(ptrToStringAnsi, schemaDisplay);
+        }
+        finally
+        {
+            Free(result);            
+        }
+    }
+    
+    /// <summary>
+    /// Get contract receive parameters in a human interpretable form.
+    ///
+    /// Receive parameters are those given to a contract entrypoint on a update call.
+    /// </summary>
+    /// <param name="schema">Module schema in hexadecimal</param>
+    /// <param name="contractName">Contract name</param>
+    /// <param name="entrypoint">Entrypoint of contract</param>
+    /// <param name="value">Receive parameters in hexadecimal</param>
+    /// <param name="schemaVersion">Optional schema version if present from module</param> 
+    /// <returns>Receive parameters in a human interpretable form</returns>
+    internal static InteropResult GetReceiveContractParameter(string schema, string contractName, string entrypoint, string value, FFIOption schemaVersion)
+    {
+        var result = IntPtr.Zero;
+        try
+        {
+            var schemaDisplay = get_receive_contract_parameter(schema, schemaVersion, contractName, entrypoint, value, ref result);
+            var ptrToStringAnsi = Marshal.PtrToStringAnsi(result);
+            return new InteropResult(ptrToStringAnsi, schemaDisplay);
+        }
+        finally
+        {
+            Free(result);    
+        }
+    }
+    
+    /// <summary>
+    /// Get contract event in a human interpretable form.
+    /// </summary>
+    /// <param name="schema">Module schema in hexadecimal</param>
+    /// <param name="contractName">Contract name</param>
+    /// <param name="value">Contract event in hexadecimal</param>
+    /// <param name="schemaVersion">Optional schema version if present from module</param>
+    /// <returns>Contract event in a human interpretable form</returns>
+    internal static InteropResult GetEventContract(string schema, string contractName, string value, FFIOption schemaVersion)
+    {
+        var result = IntPtr.Zero;
+        try
+        {
+            var schemaDisplay = get_event_contract(schema, schemaVersion, contractName, value, ref result);
+            var ptrToStringAnsi = Marshal.PtrToStringAnsi(result);
+            return new InteropResult(ptrToStringAnsi, schemaDisplay);
+        }
+        finally
+        {
+            Free(result);   
+        }
+    }
+    
+    /// <summary>
+    /// Wrap a interop result. If the interop call <see cref="Succeeded"/> the <see cref="Message"/> contains the result.
+    /// If it failed the <see cref="Message"/> will contain the error message. 
+    /// </summary>
+    /// <param name="Message">Message from interop operation.</param>
+    /// <param name="Succeeded">If the call succeeded.</param>
+    internal readonly record struct InteropResult(string? Message, bool Succeeded);
```

Comment:
Why don't we use exceptions for errors (or whatever is conventional error handling for C#)?

Reply:
Exceptions are expensive and since we either way have a result structure I think it is find to have a functional pattern here.

Reply:
> Exceptions are expensive and since we either way have a result structure I think it is find to have a functional pattern here.

That sounds like a premature optimization to me. Do we avoid exceptions in general in the app? If not, then why is this different?

My understanding is that exceptions being slow only matters if you're using it for normal control flow. And more so in C++ than languages like C#. Either way I prefer following standard conventions until you know that there's an actual problem.

If we want to use result types there's one here: https://dotnet.github.io/dotNext/api/DotNext.Result-1.html.

Reply:
I don’t want to add another dependency just to have a result wrapper :-)

I changed to exception pattern.